Issue/646

hazmat345 opened the pull request:
Fixes #646. Reorganizes the local plugin manager so key info is available in entry point processes as well as the main process.

Depends on beer-garden/brewtils#257.

TheBurchLog commented on src/app/beer_garden/local_plugins/manager.py:
```diff
@@ -7,22 +7,24 @@
 from beer_garden.local_plugins.env_help import expand_string
 from beer_garden.local_plugins.runner import ProcessRunner
 
+# This is ... complicated. See the PluginManager docstring
+lpm_proxy = None
+
```
I get why you are placing this in the Plugin Manager but it is only utilized by the App. Should we be placing this here just for a global processor variable?

hazmat345 replied:
It's only used in the App because this PR is just the groundwork for the real work that I'm doing. The main purpose of this refactor is so that other modules can now get access to *something* that let's the interact with the local plugin state.

TheBurchLog commented on src/app/beer_garden/api/entry_point.py:
```diff
@@ -179,6 +181,9 @@ def _target_wrapper(
             # Also set up plugin logging
             beer_garden.log.load_plugin_log_config()
 
+            # Local plugin manager reference
+            lpm.lpm_proxy = lpm_proxy
+
```
What is the benefit of passing forward lmp_proxy instead of connecting to it?

```python
from multiprocessing.managers import BaseManager
lpm.lpm_proxy=  BaseManager(address=('', 5000), authkey='abc')
lpm.lpm_proxy.connect()
```

hazmat345 replied:
I couldn't get connecting to it working yet :upside_down_face:

TheBurchLog commented on src/app/beer_garden/app.py:
```diff
@@ -93,6 +95,22 @@ def initialize(self):
             action=beer_garden.router.forward, name="forwarder"
         )
 
+        BaseManager.register(
+            "PluginManager",
+            callable=partial(
+                PluginManager,
+                plugin_dir=config.get("plugin.local.directory"),
+                log_dir=config.get("plugin.local.log_directory"),
+                connection_info=config.get("entry.http"),
+                username=config.get("plugin.local.auth.username"),
+                password=config.get("plugin.local.auth.password"),
+            ),
+        )
+
+        data_manager = BaseManager()
+        data_manager.start()
+        beer_garden.local_plugins.manager.lpm_proxy = data_manager.PluginManager()
```
When initializing the Base Manager, why didn't you expose it with an address object? Because if you did, then we could have any where in our code initialize this class?

```python
data_manager = BaseManager(address=('', 50000)
```

hazmat345 replied:
Yeah, I looked at that and it should work. I wasn't able to get it working yet, though.

TheBurchLog commented on src/app/beer_garden/app.py:
```diff
@@ -133,7 +151,20 @@ def handle_event(event):
-                PluginManager.instance().scan_path()
+                beer_garden.local_plugins.manager.lpm_proxy.scan_path()
+
+            elif event.name == Events.INSTANCE_INITIALIZED.name:
+                beer_garden.local_plugins.manager.lpm_proxy.handle_associate(event)
+            elif event.name == Events.INSTANCE_STARTED.name:
+                beer_garden.local_plugins.manager.lpm_proxy.handle_start(event)
+            elif event.name == Events.INSTANCE_STOPPED.name:
+                beer_garden.local_plugins.manager.lpm_proxy.handle_stop(event)
+            elif event.name == Events.SYSTEM_REMOVED.name:
+                beer_garden.local_plugins.manager.lpm_proxy.remove_system(event.payload)
+            elif event.name == Events.SYSTEM_RELOAD_REQUESTED.name:
+                beer_garden.local_plugins.manager.lpm_proxy.reload_system(event.payload)
+            elif event.name == Events.SYSTEM_RESCAN_REQUESTED.name:
+                beer_garden.local_plugins.manager.lpm_proxy.scan_path()
```
I don't know why this got pulled up. Why can't this stay with the Plugin Manager as a registered listener for events?

hazmat345 replied:
Because the Plugin Manager now lives in its own process, and that process doesn't have any plumbing to listen for events. But this is still a WIP, a lot of this logic is going to change now (the plugin manager shouldn't need to care about events for the most part anymore).

TheBurchLog commented on src/app/beer_garden/app.py:
```diff
@@ -257,10 +288,10 @@ def _shutdown(self):
             self.scheduler.shutdown(wait=False)
 
         self.logger.debug("Stopping local plugin process monitoring")
-        PluginManager.instance().stop()
+        beer_garden.local_plugins.manager.lpm_proxy.stop()
 
         self.logger.debug("Stopping local plugins")
-        PluginManager.instance().stop_all()
+        beer_garden.local_plugins.manager.lpm_proxy.stop_all()
```
Again, only this class references lpm_proxy, so it feels more like only this class should use it.

hazmat345 replied:
That's going to change

TheBurchLog reviewed (commented):
Overall I think this is really cool. I have some ideas on how we could utilize this in other avenue, but those comments will be saved for later.

My big concern is that if we are going to use this, then let's not pass the object around in memory to the sub-processes. In the sub-process, just have them create a connection to the object. We just need to reserve a super high internal port for this like 50000. Then we can really use this class as it was intended.

hazmat345 commented:
@TheBurchLog yeah, this isn't really useful yet, but I wanted you to take a look at it today. Most of your comments are things that I either explored earlier or that I wanted to add after I got the core functionality working. There's definitely still work that needs to be done here, so I'm going to hold off on addressing your comments until I have a better idea of the best way forward.

hazmat345 commented:
I'm going to go ahead and merge this. I took another (quick) look at the server / client aspect of `BaseManager` but still can't get it working. As far as I can tell there's nothing wrong with passing the proxy instance around. We can revisit if that ends up not being the case.
